src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -108,7 +108,7 @@
      * and iteration through {@link CpsThreadDump#from(CpsThreadGroup)} may occur on other threads
      * (e.g. non-blocking steps, thread dumps from the UI).
      */
-    private final NavigableMap<Integer,CpsThread> threads = new ConcurrentSkipListMap<>();
+    private final NavigableMap<Integer,CpsThread> threads = Collections.synchronizedNavigableMap(new TreeMap<>());
```

Comment:
Thanks for the suggestion, @dwnusbaum. I'm having some trouble with this. I made the `threads` field `private transient final` as you suggested, along with the corresponding changes in `readResolve` to populate the `TreeMap` from the `ConcurrentSkipListMap` on deserialization after upgrade (and vice versa on deserialization after upgrade) and the corresponding changes in `writeReplace` to copy the entries from the `ConcurrentSkipListMap` to the `TreeMap` on serialization after upgrade. But then `org.jenkinsci.plugins.workflow.cps.SerialFormTest#persistence` started failing. Debugging it with just the `private transient final` change to the `threads` field and stepping through `readResolve`, I see that marking `threads` as `transient` has made the field `null` in `readResolve`. If we were using XStream to deserialize, I think XStream would have given us the old value when reading in the `transient` field in `readReplace` but omitted the field when writing. But we are using JBoss Marshalling here. So perhaps JBoss Marshalling doesn't read in the old value if the field is marked as `transient`?

Reply:
Not sure about XStream vs. JBoss Marshalling w.r.t. `transient`, but presumably it cannot be `final` since you would need to set it to a new collection in `readResolve`.

Reply:
Correct. It was not `final` in my diff that also modified `readResolve` and `writeReplace`. I only marked it `transient final` in my diff to debug the `readResolve` method, where I found that it was unexpectedly `null` from JBoss Marshalling.

Reply:
Here is my (failing `SerialFormTest`) diff that implemented @dwnusbaum's suggestion:

```java
diff --git a/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java b/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java
index 22f0d148..7bc9fe36 100644
--- a/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java
+++ b/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java
@@ -64,6 +64,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.NavigableMap;
+import java.util.TreeMap;
 import java.util.concurrent.Callable;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentSkipListMap;
@@ -108,7 +109,12 @@ public final class CpsThreadGroup implements Serializable {
      * and iteration through {@link CpsThreadDump#from(CpsThreadGroup)} may occur on other threads
      * (e.g. non-blocking steps, thread dumps from the UI).
      */
-    private final NavigableMap<Integer,CpsThread> threads = new ConcurrentSkipListMap<>();
+    private transient NavigableMap<Integer,CpsThread> threads = new ConcurrentSkipListMap<>();
+
+    /**
+     * Persistent version of {@link #threads}.
+     */
+    private NavigableMap<Integer, CpsThread> executionThreads = new TreeMap<>();
 
     /**
      * Unique thread ID generator.
@@ -186,6 +192,14 @@ public final class CpsThreadGroup implements Serializable {
                 script.setBinding(shell.getContext());
             }
         }
+        synchronized (this) {
+            if (executionThreads == null) {
+                executionThreads = threads != null ? new TreeMap<>(threads) : new TreeMap<>();
+            }
+            if (threads == null) {
+                threads = new ConcurrentSkipListMap<>(executionThreads);
+            }
+        }
         if (paused == null) { // introduced and later removed from serial form
             paused = new AtomicBoolean(executionPaused);
         }
@@ -197,6 +211,12 @@ public final class CpsThreadGroup implements Serializable {
         pausedByQuietMode = new AtomicBoolean();
     }
 
+    private synchronized Object writeReplace() {
+        executionThreads.clear();
+        executionThreads.putAll(threads);
+        return this;
+    }
+
     @CpsVmThreadOnly
     public CpsThread addThread(@NonNull Continuable program, FlowHead head, ContextVariableSet contextVariables) {
         assertVmThread();
```

Reply:
I would not recommend `readObject`.

If you are having trouble, probably easiest to just deprecate `threads` and introduce two fresh fields.

Reply:
> Would that not create a concurrency problem, since CpsThreadGroup#saveProgram would call writeReplace and convert the field to a TreeMap, and then nothing would convert it back to a ConcurrentSkipListMap by the time CpsThreadGroup#getThread is called later on (e.g. from the onSuccess callback)?

If you use `writeObject`, then you would only modify what you write to the `ObjectOutputStream`, not `this`. You are right that using `writeReplace` would be problematic.

Reply:
(And I think Jesse means to leave the existing field non-transient, but deprecate it and only use it for one-time migration, then introduce a new transient `ConcurrentSkipListMap` field for regular use and a corresponding non-transient `TreeMap` field for serialization use only, but I am not sure.)

Reply:
Ah OK. I got this working and tested with `org.jenkinsci.plugins.workflow.support.steps.input.InputStepTest#serialForm` and `org.jenkinsci.plugins.workflow.steps.TimeoutStepTest#serialForm` as well. Since the non-`transient` map is for serialization use only, I opted to use `HashMap` rather than `TreeMap` or `LinkedHashMap`. The reason why is this has the simplest XStream representation: `<persistentThreads>` rather than `<persistentThreads class="tree-map">` or `<persistentThreads class="linked-hash-map">`. I figure this is the least likely to cause problems in the future. The entries are placed into a `ConcurrentSkipListMap` at runtime anyway, so the runtime logic should always be using the sorted map.